### PR TITLE
New code for b tagging sf

### PR DESCRIPTION
**Pull request by jedori0228**
(I closed #83 for clean page) 

Previous b tagging codes were too old, and was not fully optimized to our current analyzer, so I rewrote the b tagging codes.

- BTagCalibrationStandalone has been updated 2 years ago with some new features ([cmssw](https://twiki.cern.ch/twiki/bin/view/CMS/BTagCalibration)); e.g., eval_auto_bounds()
- BTagSFUtil removed; using MCCorrection
- MC tag efficiency is measured with the same pt binning used in the SFs
   - Before it was {20., 40., 60., 80., 100., 120., 3000.}, which is a legacy of LQAnalyzer. But the SFs are provided only up to 1000 GeV, and even the [referenced code](https://github.com/rappoccio/usercode/blob/Dev_53x/EDSHyFT/plugins/BTaggingEffAnalyzer.cc) in the LQAnalyzer used 0-1000 GeV. (@jalmond will comment if he finds the correct reference)
- Three SF application methods are added; 1a, 2a, and 2b. (https://twiki.cern.ch/twiki/bin/view/CMS/BTagSFMethods)

In initializeAnalyzer() of your analyzer,
```
  std::vector<JetTagging::Parameters> jtps;
  //==== If you want to use 1a and 2a method
  jtps.push_back( JetTagging::Parameters(JetTagging::DeepCSV, JetTagging::Medium, JetTagging::incl, JetTagging::comb) );
  //==== If you want to use 2b, which is a reshaping method
  jtps.push_back( JetTagging::Parameters(JetTagging::DeepCSV, JetTagging::Medium, JetTagging::iterativefit, JetTagging::iterativefit) );
  //==== now put temporary vector (jtps) to MCCorrection::jetTaggingPars
  //==== jetTaggingPars will be looped over and used to read csv files when initializeAnalyzerTools() is ran.
  mcCorr->SetJetTaggingParameters(jtps);
```

For this, I changed the order of functions ran in the macro;
Before, initializeAnalyzerTools() runs first and then initializeAnalyzer() runs. Now, initializeAnalyzer() first and then initializeAnalyzerTools().

Then, inside executeEvent(),

```
  int NBJets_NoSF(0), NBJets_WithSF_2a(0), NBJets_WithSF_2b(0);

  JetTagging::Parameters jtp_DeepCSV_Medium = JetTagging::Parameters(JetTagging::DeepCSV,
                                                                     JetTagging::Medium,
                                                                     JetTagging::incl, JetTagging::comb);
  JetTagging::Parameters jtp_DeepCSV_Medium_Reshape = JetTagging::Parameters(JetTagging::DeepCSV,
                                                                             JetTagging::Medium,
                                                                             JetTagging::iterativefit, JetTagging::iterativefit);

  //==== b tag SF; method 1a
  double btagWeight = mcCorr->GetBTaggingReweight_1a(myjets, jtp_DeepCSV_Medium);

  for(unsigned int i=0; i<myjets.size(); i++){
    Jet this_jet = myjets.at(i);
    double this_discr = this_jet.GetTaggerResult(JetTagging::DeepCSV);

    //==== No SF
    if( this_discr > mcCorr->GetJetTaggingCutValue(JetTagging::DeepCSV, JetTagging::Medium) ) NBJets_NoSF++;
    //==== 2a
    if( mcCorr->IsBTagged_2a(jtp_DeepCSV_Medium, this_jet) ) NBJets_WithSF_2a++;
    //==== 2b
    if( mcCorr->IsBTagged_2b(jtp_DeepCSV_Medium_Reshape, this_jet) ) NBJets_WithSF_2b++;
  }
```

### CODE REVIEW

File: AnalyzerTools/src/MCCorrection.C [hunk 1]
```diff
@@ -843,3 +852,321 @@ double MCCorrection::GetOfficialDYReweight(const vector<Gen>& gens, int sys){
...
+
+bool MCCorrection::IsBTagged_2b(JetTagging::Parameters jtp, const Jet& jet, string Syst){
+
+  double new_discr = GetScaledJetTaggerResult_2b(jtp, jet, Syst);
+  double cutValue = GetJetTaggingCutValue(jtp.j_Tagger, jtp.j_WP);
+
+  return new_discr > cutValue;
+
+}
+
```
Comment (bhoh) [inline AnalyzerTools/src/MCCorrection.C]: 2b method is also event by event weight based method.
```
double MCCorrection::GetBTaggingReweight_2b(const vector<Jet>& jets, JetTagging::Parameters jtp, string Syst){

   if(IsDATA) return 1.;

   double weight(1.);
   for(unsigned int i=0; i<jets.size(); i++){
     
     double this_SF = GetJetTaggingSF(jtp,
                                      jets.at(i).hadronFlavour(),
                                      jets.at(i).Pt(),
                                      jets.at(i).Eta(),
                                      jets.at(i).GetTaggerResult(jtp.j_Tagger),
                                      Syst );

      weight *=this_SF;
   }

   return weight;

 }
```

Comment (jedori0228) [reply AnalyzerTools/src/MCCorrection.C]: 2b) is one of the " 2) Methods **without** event reweighting " isn't it?

File: AnalyzerTools/src/MCCorrection.C [hunk 2]
```diff
@@ -843,3 +852,321 @@ double MCCorrection::GetOfficialDYReweight(const vector<Gen>& gens, int sys){
...
+      string this_key_H = tmp_tagger+"_"+this_mt_H;
+      std::map< string, BTagCalibration >::const_iterator tmp_it_H = tmp_map_BTagCalibration.find( tmp_tagger+"_"+this_key_H );
+      if(tmp_it_H==tmp_map_BTagCalibration.end()){
+        tmp_map_BTagCalibration[tmp_tagger+"_"+this_key_H] = BTagCalibration(tmp_tagger, this_mt_H, btagpath.Data()+tmp_filename);
+      }
+
+      //==== Now, contructing BTagCalibrationReader obect
+
+      std::vector<std::string> systvec= {"up", "down"};
+      if(this_mt_L=="iterativefit") systvec = {}; //TODO no systematic for iterativefit yet
```
[the remark this answers is not in the document]
Comment (jedori0228) [reply AnalyzerTools/src/MCCorrection.C]: Thanks I will update this

File: AnalyzerTools/src/MCCorrection.C (hunk 1 above)
Comment (bhoh) [reply AnalyzerTools/src/MCCorrection.C]: twiki is a bit confusing.
There are two different method in same section.
  1) One is a method **without** reweighting [presentation](https://indico.cern.ch/getFile.py/access?contribId=7&resId=0&materialId=slides&confId=162142)
  2) One is a method with reweighting [example code](https://twiki.cern.ch/twiki/bin/view/CMS/BTagShapeCalibration)

Comment (jedori0228) [reply AnalyzerTools/src/MCCorrection.C]: ohh so 1d) method from [this twiki](https://twiki.cern.ch/twiki/bin/view/CMS/BTagSFMethods#1d_Event_reweighting_using_discr) is what we should do with that SFs?

Comment (bhoh) [reply AnalyzerTools/src/MCCorrection.C]: I think you are right!

Comment (jedori0228) [reply AnalyzerTools/src/MCCorrection.C]: thanks, I will update this too

**Comment by jedori0228** [comment]
I made some validation plots with event selection as follows;

- Two tight muons with m(ll) > 110 GeV
- at least two AK4 jets (lepton veto jets)
- MET > 30 GeV

1. DeepCSV discriminants with and without SF (using 1d method) : [without SF](https://jskim.web.cern.ch/jskim/HNWR_13TeV/Run2Legacy_v4__Default/SKFlatValidation/2016/InclusiveDY/POG_SingleMuon_mllgt110_TwoJets_METgt30_OS//Jet_0_DeepCSV_POG_SingleMuon_mllgt110_TwoJets_METgt30_OS.pdf) / [with SF, 1d](https://jskim.web.cern.ch/jskim/HNWR_13TeV/Run2Legacy_v4__Default/SKFlatValidation/2016/InclusiveDY/POG_SingleMuon_mllgt110_TwoJets_METgt30_OS//Jet_0_DeepCSV_Scaled_POG_SingleMuon_mllgt110_TwoJets_METgt30_OS.pdf)

2. Number of b-tagged jets : [without SF](https://jskim.web.cern.ch/jskim/HNWR_13TeV/Run2Legacy_v4__Default/SKFlatValidation/2016/InclusiveDY/POG_SingleMuon_mllgt110_TwoJets_METgt30_OS//NBJets_NoSF_POG_SingleMuon_mllgt110_TwoJets_METgt30_OS.pdf), [with SF, 1a](https://jskim.web.cern.ch/jskim/HNWR_13TeV/Run2Legacy_v4__Default/SKFlatValidation/2016/InclusiveDY/POG_SingleMuon_mllgt110_TwoJets_METgt30_OS//NBJets_WithSF_1a_POG_SingleMuon_mllgt110_TwoJets_METgt30_OS.pdf), [with SF, 1d](https://jskim.web.cern.ch/jskim/HNWR_13TeV/Run2Legacy_v4__Default/SKFlatValidation/2016/InclusiveDY/POG_SingleMuon_mllgt110_TwoJets_METgt30_OS//NBJets_WithSF_1d_POG_SingleMuon_mllgt110_TwoJets_METgt30_OS.pdf), [with SF, 2a](https://jskim.web.cern.ch/jskim/HNWR_13TeV/Run2Legacy_v4__Default/SKFlatValidation/2016/InclusiveDY/POG_SingleMuon_mllgt110_TwoJets_METgt30_OS//NBJets_WithSF_2a_POG_SingleMuon_mllgt110_TwoJets_METgt30_OS.pdf)

File: AnalyzerTools/src/MCCorrection.C [hunk 3]
```diff
@@ -1079,6 +1088,34 @@ double MCCorrection::GetBTaggingReweight_1a(const vector<Jet>& jets, JetTagging:
...
+  double rew(1.);
+  for(unsigned int i=0; i<jets.size(); i++){
+    double this_SF = GetJetTaggingSF(jtp,
+                                     jets.at(i).hadronFlavour(),
+                                     jets.at(i).Pt(),
+                                     jets.at(i).Eta(),
+                                     jets.at(i).GetTaggerResult(jtp.j_Tagger),
+                                     Syst );
+    rew *= this_SF;
+  }
```
Comment (bhoh) [inline AnalyzerTools/src/MCCorrection.C]:
```
    double rew(1.);
    for(unsigned int i=0; i<jets.size(); i++){
      int abs_hadFlavour = abs(jets.at(i).hadronFlavour());
      TString tmp_Syst(Syst);
      if(tmp_Syst.Contains(TRegexp("cferr[1-2]$")) && abs_hadFlavour!=4 ||
          tmp_Syst.Contains(TRegexp("jes$")) && abs_hadFlavour==4 ||
          tmp_Syst.Contains(TRegexp("hf$")) && abs_hadFlavour>=4 ||
          tmp_Syst.Contains(TRegexp("lf$")) && abs_hadFlavour<5 ||
          tmp_Syst.Contains(TRegexp("lfstat[1-2]$")) && abs_hadFlavour>=4 ||
          tmp_Syst.Contains(TRegexp("hfstat[1-2]$")) && abs_hadFlavour<5
         ){
        tmp_Syst = "central";
      }
      double this_SF = GetJetTaggingSF(jtp,
                                        tmp_hadFlavour,
                                        jets.at(i).Pt(),
                                        jets.at(i).Eta(),
                                        jets.at(i).GetTaggerResult(jtp.j_Tagger),
                                        string(tmp_Syst.Data()));
      rew *= this_SF;
    }
```

Comment (jedori0228) [reply AnalyzerTools/src/MCCorrection.C]: Thank you, I updated this

**Pull request merged**